book/core-concepts/windows.md

```diff
@@ -0,0 +1,152 @@
+|00:00|00:01|00:02|00:03|00:04|00:05|00:06|00:07|00:08|00:09|00:10|00:11
+|-----------------|-----------------|-----------------|-----------------|
+     window 1          window 2          window 3          window 4
+   [00:00,00:03)     [00:03,00:06)     [00:06,00:09)     [00:09,00:12)
+```
+
+Each window above is indicated by an interval such as `[00:00,00:03)`, indicating "the window containing all times equal to or greater than `00:00` and less than `00:03`".
+
+Incoming messages will be placed in a window based on the event time of the message. In the near future, you will be able to derive this event time from each input message that arrives at a Wallaroo source. For now, it is always the ingest time for the input message.
+
+So if we had the following four messages (min:sec:ms):
+```
+<id: 1, event_time: 00:04:043>
+<id: 2, event_time: 00:01:022>
+<id: 3, event_time: 00:06:000>
+<id: 4, event_time: 00:09:055>
+```
+
+They would be placed as follows:
+```
+<id: 1, event_time: 00:04:043> --> window 2 [00:03,00:06)
+<id: 2, event_time: 00:01:022> --> window 1 [00:00,00:03)
+<id: 3, event_time: 00:06:000> --> window 3 [00:06,00:09)
+<id: 4, event_time: 00:05:655> --> window 2 [00:03,00:06)
+```
+
+Notice that it doesn't matter which order the messages arrive in. They will be placed in the correct windows based on event time.
+
+However, if there were no limit to how late messages can be, then we would have to keep around the state for each window indefinitely. This can add up quickly in a high-volume scenario. As a result, we currently throw away window data once the window output is triggered (i.e. the aggregation `output()` method is called and an output is produced for the window). This means that from then on, late messages are dropped on the floor. In future versions of Wallaroo, we will provide more fine-grained policies so that you can choose other options for late messages. In the meantime, the one knob we provide is a `delay` on triggering a window. For example:
+
+```python
+    (inputs
+        .to(wallaroo.range_windows(wallaroo.seconds(3))
+            .with_delay(wallaroo.seconds(3))
+            .over(MySumAgg))
+        .to_sink(sink_config))
+```
+
+This says that we delay triggering each window for 3 seconds. Think of the delay knob as setting either (1) an estimation of the maximum lateness for messages or (2) the lateness threshold beyond which you no longer care about messages. The knob is essentially trading off between latency (longer delays means longer waits for window outputs) and completeness (shorter delays potentially mean more ignored messages). 
+
+IMPORTANT NOTE: If your application requires a guarantee that no messages are ever ignored, then it's not currently safe to use the Wallaroo windowing API unless you are certain that no messages will ever be later than your delay setting. However, as mentioned above, we will be adding more fine-grained policy control for late message in future versions.
```

Comment:
I recommend adding a note about it being possible to do in the mean time using custom state

Reply:
I would rather we don't make this kind of recommendation.  Using custom state to try to solve this problem involves all kinds of problems of its own, including the possibility of effective message loss if you don't manage your own trigger pipeline correctly.  And if you don't properly evict outdated windows, there's also OOM killing.